presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/users/UserEditViewModel.kt

```diff
@@ -1,61 +1,93 @@
 import com.jakewharton.rxrelay2.PublishRelay
-import io.reactivex.Observable
-import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.disposables.Disposable
 
-class UserEditViewModel(userEditModel: UserEditModel,
-                        private val router: UiRouter) {
+class UserEditViewModel(
+        private val userEditModel: UserEditModel,
+        private val router: UiRouter,
+        private val firstNameRelay: BehaviorRelay<String> = BehaviorRelay.create(),
+        private val lastNameRelay: BehaviorRelay<String> = BehaviorRelay.create(),
+        private val savedUserRelay: PublishRelay<User> = PublishRelay.create(),
+        ecoDisposableProperty: EcoDisposableProperty = ecoDisposable(null),
+        notifiableObservable: NotifiableObservable = BaseNotifiableObservable())
+    : NotifiableObservable by notifiableObservable {
 
-    val firstNameError = ObservableField<String>("")
-    val lastNameError = ObservableField<String>("")
+    init {
+        setDelegator(this)
+    }
 
-    private val compositeDisposable = CompositeDisposable()
+    /* symbol meaning:
+    >> : bind direction left to right, future changes are propagated
+    << : bind direction right to left, future changes are propagated
+    >| : bind direction left to right, future changes are not propagated
+    |< : bind direction right to left, future changes are not propagated
+     */
 
-    private val clickRelay = PublishRelay.create<User>()
-    private val firstNameRelay = BehaviorRelay.create<String>()
-    private val lastNameRelay = BehaviorRelay.create<String>()
+    // view |< >> view model
+    var firstName: String by distinctObservable(
+            initialValue = "",
+            onChange = firstNameRelay::accept)
 
-    init {
-        compositeDisposable.add(userEditModel.results(clickRelay, firstNameRelay, lastNameRelay)
-                .subscribe(this::handleResult, this::handleError))
-    }
+    // view |< >> view model
+    var lastName: String by distinctObservable(
+            initialValue = "",
+            onChange = lastNameRelay::accept)
 
-    fun setSaveClicks(obs: Observable<Unit>) {
-        compositeDisposable.add(obs
-                .map { User(firstNameRelay.value, lastNameRelay.value) }
-                .subscribe(clickRelay))
-    }
+    // view << view model
+    @get:Bindable
+    var firstNameError: String by bindable("")
+
+    // view << view model
+    @get:Bindable
+    var lastNameError: String by bindable("")
+
+    private var disposable: Disposable? by ecoDisposableProperty
+    private var savedUser: User by distinctObservable(
+            initialValue = User(firstName, lastName),
+            onChange = savedUserRelay::accept,
+            onChangeOnInit = false)
 
-    fun setFirstNameChanged(obs: Observable<String>) {
-        compositeDisposable.add(obs.subscribe(firstNameRelay))
+    // view >> view model
+    fun saveUser() {
+        savedUser = User(firstName, lastName)
```

Comment:
I'm not in love with `savedUser`.  It isn't clear here via naming that we're performing an action